src/Tools/DependencyManagement/ObjectContainer.php

```diff
@@ -0,0 +1,169 @@
+<?php
+/**
+ * ObjectContainer class file.
+ *
+ * @package Automattic\WooCommerce\Tools\DependencyManagement
+ */
+
+namespace Automattic\WooCommerce\Tools\DependencyManagement;
+
+use \League\Container\Container;
+use \Automattic\WooCommerce\Tools\DependencyManagement\ReflectionContainer as WooReflectionContainer;
+
+/**
+ * Dependency injection container for WooCommerce.
+ *
+ * This class uses PHP League's `Container` under the hood. A separate container class is used to
+ * insulate the codebase from the exact DI engine used, and to encapsulate all the class and service provider registrations.
+ *
+ * Usage:
+ *
+ * - To register new class resolutions:
+ *   - Do it in the `register_classes` method, using the `container` and `reflection_container` class properties.
+ *     In particular, use `reflection_container->share` to register autowiring singletons.
+ *     No extra action is needed to use non-singleton autowiring.
+ *   - Or alternatively, create a new service provider inside the `ServiceProviders` folder and it will be
+ *     automatically registered. Service providers are recommended when instantiating the class is a complex
+ *     or slow process.
+ *
+ * - To resolve classes to instances:
+ *   - Ideally you just add your dependency as a typed argument in the class constructor, and it gets resolved
+ *     automatically (via autowiring) when the class is instantiated via the container. This is the preferred approach.
+ *   - If the above is not possible, you can add an argument of type `ObjectContainer` to the class constructor,
+ *     and then use the `get` method of the supplied instance whenever needed.
+ *   - For code outside classes (inside standalone functions) or in static class methods, use
+ *    `ObjectContainer::get_instance_of`.
+ *
+ * Note: autowiring means the ability to automatically resolve a class name to an instance of that class via
+ *       `get(TheClass::class)` or by adding a typed argument to class constructors.
+ */
+final class ObjectContainer {
+
+	/**
+	 *  The singleton instance of ObjectContainer.
+	 *
+	 * @var ObjectContainer
+	 */
+	private static $instance = null;
```

Comment:
If we are going to discourage/obsolete the usage of old code whenever possible, I think that should include the `WC()` function, therefore I wouldn't add anything new to it. The container is a special case where it's appropriate to keep it as singleton (but it's also registered so it can be passed as an argument constructor as well).